Fix error codes from various RPCs

jnewbery opened the pull request:
(Completes #9707 and #9842)

I've changed the qa test code to properly assert error codes and messages for failed RPC calls. That's revealed a few RPCs which were returning incorrect or misleading error codes. This PR fixes all of those cases.

It also adds commenting around the definition of the JSON RPC error constants to warn against using RPC_INVALID_REQUEST and RPC_METHOD_NOT_FOUND for application-layer errors.

jnewbery committed:
Return correct error codes in setban().

The setban() RPC was returning misleading or incorrect error
codes (for example RPC_CLIENT_NODE_ALREADY_ADDED when an invalid IP
address was entered). This commit fixes those error codes:

- RPC_CLIENT_INVALID_IP_OR_SUBNET should be returned if the client
  enters an invalid IP address or subnet.

This commit also updates the test cases to explicitly test the error code.

This commit also adds a testcase for trying to setban on an invalid subnet.

jnewbery committed:
Return correct error codes in bumpfee().

The bumpfee() RPC was returning misleading or incorrect error codes
(for example RPC_INVALID_ADDRESS_OR_KEY when the transaction was not
BIP125 replacable). This commit fixes those error codes:

- RPC_INVALID_ADDRESS_OR_KEY if an invalid address was provided:
    - Invalid change address given
- RPC_INVALID_PARAMETER if a single (non-address/key) parameter is incorrect
    - confTarget and totalFee options should not both be set.
    - Invalid confTarget
    - Insufficient totalFee (cannot be less than required fee)
- RPC_WALLET_ERROR for any other error
    - Transaction has descendants in the wallet
    - Transaction has descendants in the mempool
    - Transaction has been mined, or is conflicted with a mined transaction
    - Transaction is not BIP 125 replaceable
    - Transaction has already been bumped
    - Transaction contains inputs that don't belong to the wallet
    - Transaction has multiple change outputs
    - Transaction does not have a change output
    - Fee is higher than maxTxFee
    - New fee rate is less than the minimum fee rate
    - Change output is too small.

This commit also updates the test cases to explicitly test the error code.

jnewbery committed:
Return correct error codes in fundrawtransaction().

The fundrawtransaction() RPC was returning misleading or incorrect error
codes (for example RPC_INTERNAL_ERROR when funding the transaction
failed). This commit fixes those error codes:

- RPC_INTERNAL_ERROR should not be returned for application-level
errors, only for genuine internal errors such as corrupted data.

That error code has been replaced with RPC_WALLET_ERROR.

This commit also updates the test cases to explicitly test the error code.

jnewbery committed:
Return correct error codes in removeprunedfunds().

The removeprunedfunds() RPC was returning misleading or incorrect error
codes (for example RPC_INTERNAL_ERROR when the transaction was
not found in the wallet). This commit fixes those error codes:

- RPC_INTERNAL_ERROR should not be returned for application-level
errors, only for genuine internal errors such as corrupted data.

This error code has been replaced with RPC_WALLET_ERROR.

This commit also updates the test cases to explicitly test the error code.

jnewbery committed:
Return correct error codes in blockchain.cpp.

RPCs in blockchain.cpp were returning misleading or incorrect error
codes (for example getblock() returning RPC_INTERNAL_ERROR when the
block had been pruned). This commit fixes those error codes:

- RPC_INTERNAL_ERROR should not be returned for application-level
  errors, only for genuine internal errors such as corrupted data.
- RPC_METHOD_NOT_FOUND should not be returned in response to a
  JSON request for an existing method.

Those error codes have been replaced with RPC_MISC_ERROR or
RPC_INVALID_PARAMETER as appropriate.

jnewbery committed:
Add commenting around JSON error codes

RPC_INVALID_REQUEST and RPC_METHOD_NOT_FOUND are mapped internally to
HTTP error codes and should not be used for application-layer errors.
This commit adds commenting around those definitions to warn not to use
them for application errors.

ryanofsky commented on qa/rpc-tests/bumpfee.py:
```diff
@@ -128,7 +128,7 @@ def test_segwit_bumpfee_succeeds(rbf_node, dest_address):
     # cannot replace a non RBF transaction (from node which did not enable RBF)
     not_rbfid = create_fund_sign_send(peer_node, {dest_address: 0.00090000})
-    assert_raises_message(JSONRPCException, "not BIP 125 replaceable", peer_node.bumpfee, not_rbfid)
+    assert_raises_jsonrpc(-4, "not BIP 125 replaceable", peer_node.bumpfee, not_rbfid)
```
Is there a plan to switch from integer literals to constants in a later PR? Seems like it would be nice to say `RPC_WALLET_ERROR` instead of `-4`.

laanwj replied:
Yes that would be nice, but is orthogonal to this change, so if you want to do that it'd indeed be a different PR.

jnewbery replied:
yep, I'll put that in the todo list. I'm working on a primitives.py for bitcoin messages, datastructures and constants. Seems like a good place for the RPC error constants to live.

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -2753,33 +2753,33 @@ UniValue bumpfee(const JSONRPCRequest& request)
 
     if (pwalletMain->HasWalletSpend(hash)) {
-        throw JSONRPCError(RPC_MISC_ERROR, "Transaction has descendants in the wallet");
+        throw JSONRPCError(RPC_WALLET_ERROR, "Transaction has descendants in the wallet");
```
In cases where bumpfee is called erroneously with a transaction that shouldn't be bumped, invalid parameter seems more apt than wallet error.

Probably the most important thing you want to know when you get an error message from an API is whether the error is caused by something that you control and need to fix, or whether it's caused by something out of your control (like an implementation bug, or corrupt data, or a temporary failure). RPC_INVALID_PARAMETER clearly tells you that you need to fix something, while RPC_WALLET_ERROR is vague but suggests that something's wrong with the wallet.

jnewbery replied:
> Probably the most important thing you want to know when you get an error message from an API is whether the error is caused by something that you control and need to fix, or whether it's caused by something out of your control

Nicely put. I'll update this to RPC_INVALID_PARAMETER

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -338,11 +338,11 @@ UniValue removeprunedfunds(const JSONRPCRequest& request)
 
     if(vHashOut.empty()) {
-        throw JSONRPCError(RPC_INTERNAL_ERROR, "Transaction does not exist in wallet.");
+        throw JSONRPCError(RPC_WALLET_ERROR, "Transaction does not exist in wallet.");
```
Maybe invalid parameter here as well.

jnewbery replied:
ACK

ryanofsky commented on src/wallet/rpcdump.cpp:
```diff
@@ -338,11 +338,11 @@ UniValue removeprunedfunds(const JSONRPCRequest& request)
 
     if(pwalletMain->ZapSelectTx(vHash, vHashOut) != DB_LOAD_OK) {
-        throw JSONRPCError(RPC_INTERNAL_ERROR, "Could not properly delete the transaction.");
+        throw JSONRPCError(RPC_WALLET_ERROR, "Could not properly delete the transaction.");
```
ZapSelect returns a DBErrors enum value which seems like it could be mapped to an RPC error (would also be nice to include in the error string).

jnewbery replied:
I agree we could do better, but let's leave that for a future PR.

I think this is strictly an improvement because currently the API is returning RPC_INTERNAL_ERROR even in cases where there's no internal error.

ryanofsky commented on src/rpc/blockchain.cpp:
```diff
@@ -829,7 +834,7 @@ UniValue pruneblockchain(const JSONRPCRequest& request)
 
     if (!fPruneMode)
-        throw JSONRPCError(RPC_METHOD_NOT_FOUND, "Cannot prune blocks because node is not in prune mode.");
+        throw JSONRPCError(RPC_MISC_ERROR, "Cannot prune blocks because node is not in prune mode.");
```
Some of these MISC_ERROR cases make it seem like we could benefit from an new RPC_INVALID_STATE value. (Blockchain too short is another case below.)

jnewbery replied:
Sounds sensible, but again let's leave that for a future PR.

laanwj replied:
Yes, there is no reason why we couldn't add new status values. There's no need to make everything fit in the arbitrary bag of codes that exists now, which has been there unchanged for a long while.

ryanofsky reviewed (commented):
utACK b34faf156d49705faa335641cfbfc0efd60bb781, definitely an improvement.

jnewbery commented:
Squashed @ryanofsky's nits and rebased.
Old tip: https://github.com/bitcoin/bitcoin/pull/9853/commits/5158a1e2a2587f9eba04d0f2a81744a9ecbe766d / https://github.com/jnewbery/bitcoin/tree/pr9853.1
New tip: https://github.com/bitcoin/bitcoin/pull/9853/commits/3ff485e32f7a1d068e799d323d3531ede22873f1 / https://github.com/jnewbery/bitcoin/tree/pr9853.2

laanwj commented:
utACK 3ff485e

MarcoFalke commented:
Concept ACK.

Style Nit: In blockchain.cpp you are missing braces around the
multiline if block

Also, I was wondering if this requires a mention in the release notes...

laanwj commented:
> Also, I was wondering if this requires a mention in the release notes...

It does.

ryanofsky commented:
utACK 3ff485e32f7a1d068e799d323d3531ede22873f1

Confirmed only minor changes since previous ACK.

jnewbery committed:
Update release notes to include RPC error code changes.

jnewbery commented:
Release notes updated. Otherwise no change.
